akka-stream-tests/src/test/scala/akka/stream/io/TlsSpec.scala

```diff
@@ -0,0 +1,312 @@
+      def inputs = List(send("hello"), NegotiateNewSession.default, send("world"))
+      def output = ByteString("helloNEWSESSIONworld")
+    }
+
+    // difference is that the RHS engine will now receive the handshake while trying to send
+    object SessionRenegotiationByReceiver extends PayloadScenario {
+      val str = "abcdef" * 100
+      def inputs = str.map(send) ++ Seq(NegotiateNewSession.default) ++ "hello world".map(send)
+      def output = ByteString(str + "NEWSESSIONhello world")
+    }
+
+    val logCipherSuite = Flow[SslTlsInbound]
+      .map {
+        var session: SSLSession = null
+        def setSession(s: SSLSession) = {
+          session = s
+          system.log.debug(s"new session: $session (${session.getId mkString ","})")
+        }
+
+        {
+          case SessionTruncated ⇒ SendBytes(ByteString("TRUNCATED"))
+          case SessionBytes(s, b) if s != session ⇒
+            setSession(s)
+            SendBytes(ByteString(s.getCipherSuite) ++ b)
+          case SessionBytes(s, b) ⇒ SendBytes(b)
+        }
+      }
+
+    object SessionRenegotiationFirstOne extends PayloadScenario {
+      override def flow = logCipherSuite
+      def inputs = NegotiateNewSession.withCipherSuites("TLS_RSA_WITH_AES_128_CBC_SHA") :: send("hello") :: Nil
+      def output = ByteString("TLS_RSA_WITH_AES_128_CBC_SHAhello")
+    }
+
+    object SessionRenegotiationFirstTwo extends PayloadScenario {
+      override def flow = logCipherSuite
+      def inputs = NegotiateNewSession.withCipherSuites("TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA") :: send("hello") :: Nil
+      def output = ByteString("TLS_ECDHE_RSA_WITH_AES_128_CBC_SHAhello")
+    }
+
+    val scenarios =
+      Seq(
+        SingleBytes,
+        MediumMessages,
+        LargeMessages,
+        EmptyBytesFirst,
+        EmptyBytesInTheMiddle,
+        EmptyBytesLast,
+        ImpatientRHS,
+        SessionRenegotiationBySender,
+        SessionRenegotiationByReceiver,
+        SessionRenegotiationFirstOne,
+        SessionRenegotiationFirstTwo)
+
+    for {
+      commPattern ← communicationPatterns
+      scenario ← scenarios
+    } {
+      s"work in mode ${commPattern.name} while sending ${scenario.name}" in {
+        val onRHS = debug.via(scenario.flow)
+        val f =
+          Source(scenario.inputs)
+            .via(commPattern.flow(scenario.leftClosing, scenario.rightClosing, onRHS))
+            .via(debug)
+            .collect { case SessionBytes(_, b) ⇒ b }
+            .scan(ByteString.empty)(_ ++ _)
+            .transform(() ⇒ new Timeout)
+            .dropWhile(_.size < scenario.output.size)
+            .runWith(Sink.head)
+
+        Await.result(f, 3.seconds).utf8String should be(scenario.output.utf8String)
+      }
+    }
+
+  }
+
+  // TODO: Add parallel
+  // TODO: Add chained parallel, loopback
+  // TODO: Add inverted chains (client-server-server-client and client-server-client-server and server-client-client-server)
```

Comment:
that's what I said, too. I just tried to explain the TODOs because they are terse and was meant to me as a reminder.